### PR TITLE
Python bindings improvements

### LINKED ISSUE #762: Python bindings for new R/U/G types
In the previous weeks, a lot of new types have been added to the C++ library. It's time to make them available in our Python bindings:
- [x] Implement Python bindings for all transformation classes (R/U/G);
- [x] Implement Python bindings for all second-quantized operators (R/U/G one- and two-electron);
- [x] Implement Python bindings for all density matrices (orbital, spin-resolved, G);
- [x] Implement Python bindings for all spinor/spin-orbital basis (R/U/G);

Since we're working with a hierarchy of classes/interfaces, it will be important to check how we can implement this **without duplicating much code**.

### PR DESCRIPTION
**Pull request by lelemmen**
**Short description**
In this PR, I'll tackle the improvement of our current Python bindings, in order to make it close #762.

**Don't forget:**
- [ ] if new files are created: update the Doxygen file, the collective gqcp.hpp header and the CMake files

### CODE REVIEW

**Comment by lelemmen** [comment]
@xdvriend I didn't realize the addition of the extra density matrices and operators entailed this little work. Otherwise I would have done it in the previous PR.

File: gqcpy/src/DensityMatrix/G1DM_bindings.cpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    // Define the Python class for `G1DM`.
+    py::class_<G1DM<double>> py_G1DM_d {module, "G1DM_d", "A type used to represent a one-electron general(ized) density matrix, i.e. the full spinor two-component one-electron density matrix."};
+
+
+    // Expose the `BasisTransformable` API to the Python class.
+    bindBasisTransformableInterface(py_G1DM_d);
+}
+
+
+}  // namespace gqcpy
```
Comment (xdvriend) [inline gqcpy/src/DensityMatrix/G1DM_bindings.cpp:48]:
```suggestion
}  // namespace gqcpy

```

File: gqcpy/src/DensityMatrix/SpinDensity1DM_bindings.cpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    // Define the Python class for `SpinDensity1DM`.
+    py::class_<SpinDensity1DM<double>> py_SpinDensity1DM_d {module, "SpinDensity1DM_d", "A type used to represent a one-electron spin-density density matrix, i.e. the alpha density matrix minus the beta density matrix."};
+
+
+    // Expose the `BasisTransformable` API to the Python class.
+    bindBasisTransformableInterface(py_SpinDensity1DM_d);
+}
+
+
+}  // namespace gqcpy
```
Comment (xdvriend) [inline gqcpy/src/DensityMatrix/SpinDensity1DM_bindings.cpp:48]:
```suggestion
}  // namespace gqcpy

```

File: gqcpy/src/DensityMatrix/SpinResolved1DMComponent_bindings.cpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    // Define the Python class for `SpinResolved1DMComponent`.
+    py::class_<SpinResolved1DMComponent<double>> py_SpinResolved1DMComponent_d {module, "SpinResolved1DMComponent_d", "One of the spin components of a `SpinResolved1DM`."};
+
+
+    // Expose the `BasisTransformable` API to the Python class.
+    bindBasisTransformableInterface(py_SpinResolved1DMComponent_d);
+}
+
+
+}  // namespace gqcpy
```
Comment (xdvriend) [inline gqcpy/src/DensityMatrix/SpinResolved1DMComponent_bindings.cpp:48]:
```suggestion
}  // namespace gqcpy

```

File: gqcpy/src/DensityMatrix/SpinResolved2DMComponent_bindings.cpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    // Define the Python class for `SpinResolved2DMComponent`.
+    py::class_<SpinResolved2DMComponent<double>> py_SpinResolved2DMComponent_d {module, "SpinResolved2DMComponent_d", "One of the spin components of a `SpinResolved2DM`."};
+
+
+    // Expose the `Simple2DM` API to the Python class.
+    bindSimple2DMInterface(py_SpinResolved2DMComponent_d);
+}
+
+
+}  // namespace gqcpy
```
Comment (xdvriend) [inline gqcpy/src/DensityMatrix/SpinResolved2DMComponent_bindings.cpp:48]:
```suggestion
}  // namespace gqcpy

```

**Review by xdvriend — APPROVED** [review]
Good to go after some very minor adjustments in my opinion!

File: gqcpy/src/DensityMatrix/G2DM_bindings.cpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    // Define the Python class for `G2DM`.
+    py::class_<G2DM<double>> py_G2DM_d {module, "G2DM_d", "A type used to represent a two-electron general(ized) density matrix, i.e. the full spinor two-component two-electron density matrix."};
+
+
+    // Expose the `Simple2DM` API to the Python class.
+    bindSimple2DMInterface(py_G2DM_d);
+}
+
+
+}  // namespace gqcpy
```
Comment (lelemmen) [inline gqcpy/src/DensityMatrix/G2DM_bindings.cpp:48]:
```suggestion
}  // namespace gqcpy

```

**Pull request merged**